### PR TITLE
Add offline backup & restore to immuadmin

### LINKED ISSUE #186: Add offline backup & restore to immuadmin

### PR DESCRIPTION
**Pull request by padurean**
closes #186

### CODE REVIEW

**Comment by padurean** [comment]
@mmeloni I've kept 2 versions for backup and restore:
- offline (default) - no comm between server and client (possible since `immuadmin` can only run on the same machine as `immudb` server)
   - **start** and **stop** of the server can be
      - automatically (default) - works only if `immudb` is daemonized
      - manually - by the user; command has to be run with `--manual-stop-start` flag
- online - has to be enabled with the `--online` flag
   - server handles all the backup or restore process
   - server will not be actually stopped or restarted, instead stores will be set to nil and new ones will be open

Please let me know what you think and whether we should keep both or just the offline version.
Thanks.

File: pkg/fs/copy.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+package fs
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
```
Comment (mmeloni) [inline pkg/fs/copy.go:26]: could you please add a // todo Michele at CopyExecInOsDefault unix and window methods? In this way I will replace same function with this to increase consistency code

File: pkg/store/store.go [hunk 2]
```diff
@@ -457,6 +458,12 @@ func (t *Store) ZAdd(zaddOpts schema.ZAddOptions, options ...WriteOption) (index
 	return index, err
 }
 
+func (t *Store) FlushToDisk() {
+	defer t.tree.Unlock()
```
Comment (mmeloni) [inline pkg/store/store.go:462]: Maybe this is not safe. We can have pending writes in worker
try to add https://github.com/codenotary/immudb/blob/master/pkg/store/store.go#L72
Or maybe a better solution can be to use an approach similar to stop method
https://github.com/codenotary/immudb/blob/master/pkg/store/store.go#L65 in which we have 
```
t.wg.Wait()
t.tree.Close()
```

**Comment by mmeloni** [comment]
> * offline (default) - no comm between server and client (possible since `immuadmin` can only run on the same machine as `immudb` server)
>   
>   * **start** and **stop** of the server can be
>     
>     * automatically (default) - works only if `immudb` is daemonized
>     * manually - by the user; command has to be run with `--manual-stop-start` flag
> * online - has to be enabled with the `--online` flag
>   
>   * server handles all the backup or restore process
>   * server will not be actually stopped or restarted, instead stores will be set to nil and new ones will be open
> 
> Please let me know what you think and whether we should keep both or just the offline version.
> Thanks.

In my opinion for the beginning is enough to use a manual stop. It's tricky to manage immudb  backup under heavy client loads. However if immu restore and backup will works under load( I will check when you fix restore backup under apitest suite run) I will accept PR

File: cmd/immuadmin/command/backup.go [hunk 3]
```diff
@@ -53,3 +63,312 @@ func (cl *commandline) dumpToFile(cmd *cobra.Command) {
...
+			response, err := cl.immuClient.Backup(ctx, uncompressed)
+			if err != nil {
+				c.QuitWithUserError(err)
+			}
+			fmt.Printf("Database backup created: %s\n", response.GetMessage())
+			return nil
+		},
+		Args: cobra.NoArgs,
+	}
+	ccmd.Flags().String("dbdir", defaultDbDir, fmt.Sprintf("path to the server database directory to backup (default %s)", defaultDbDir))
```
[the remark this answers is not in the document]
Comment (padurean) [reply cmd/immuadmin/command/backup.go:139]: @mmeloni In _**online**_ mode this flag is ignored (a warning is shown to the user stating this), but in _**offline**_ mode the entire backup/restore happens on the client so the server db dir is not known; more so, the server can even be stopped manually by the user before backup.

**Comment by padurean** [comment]
@mmeloni The archive temp folder is actually removed. The _**db_bkp_before_restore_..**_. folder are is a safety backup of the `db` folder _**before**_ restore. So that if one changes his minds and want to revert the restore, he can stop the server, delete the **db** folder and rename back the _**db_bkp_before_restore_..**_ to _**db**_. This way restore would be reverted/canceled.

**Comment by padurean** [comment]
@mmeloni the `immudb` crash on `restore` with error `Unlock of unlocked RWMutex` has now been fixed.

**Comment by padurean** [comment]
@mmeloni i agree that _**offline**_ method is enough, especially since `immuadmin` can only be run on the same machine as the `immudb` server. The only advantage of the _**online**_ method is that one doesn't have to stop the server, while with the **_offline_** method one <ins>**has**</ins> to manually stop and restart server <ins>**if**</ins> `immudb` is not running as a service.
But if you think **_online_** is unsafe or not working fine, we can remove it.

**Comment by mmeloni** [comment]
Thanks @padurean, 
unfortunately we still have an error:

./immuadmin backup --online 

```
goroutine 166 [running]:
github.com/dgraph-io/badger/v2.(*Txn).commitPrecheck(...)
	/home/falce/vchain/immudb/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.0-20200408100755-2e708d968e94/txn.go:515
github.com/dgraph-io/badger/v2.(*Txn).CommitWith(0xc0003d1d00, 0xc010740960)
	/home/falce/vchain/immudb/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.0-20200408100755-2e708d968e94/txn.go:586 +0x2f4
github.com/dgraph-io/badger/v2.(*WriteBatch).commit(0xc010671fc0, 0x0, 0x0)
	/home/falce/vchain/immudb/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.0-20200408100755-2e708d968e94/batch.go:144 +0x9e
github.com/dgraph-io/badger/v2.(*WriteBatch).Flush(0xc010671fc0, 0x0, 0x0)
	/home/falce/vchain/immudb/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.0-20200408100755-2e708d968e94/batch.go:155 +0x45
github.com/codenotary/immudb/pkg/store.(*treeStore).flush.func1(0xc0107d843f, 0xc010671fc0, 0xc000018000)
	/home/falce/vchain/immudb/src/pkg/store/treestore.go:311 +0x6a
github.com/codenotary/immudb/pkg/store.(*treeStore).flush(0xc000018000)
	/home/falce/vchain/immudb/src/pkg/store/treestore.go:356 +0x1ea
github.com/codenotary/immudb/pkg/store.(*Store).FlushToDisk(0xc0003b4050)
	/home/falce/vchain/immudb/src/pkg/store/store.go:465 +0x9d
github.com/codenotary/immudb/pkg/server.(*ImmuServer).Backup(0xc0003d25a0, 0xf42060, 0xc0107864e0, 0xc010786510, 0x0, 0x0, 0x0)
	/home/falce/vchain/immudb/src/pkg/server/server.go:542 +0x130
github.com/codenotary/immudb/pkg/api/schema._ImmuService_Backup_Handler.func1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xc00025cb38, 0xc01076c601, 0xc0002207d8, 0xb70681)
	/home/falce/vchain/immudb/src/pkg/api/schema/schema.pb.go:2880 +0x86
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xd85900, 0xc010786510, 0x0, 0x0)
	/home/falce/vchain/immudb/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.0/chain.go:31 +0x114
github.com/codenotary/immudb/pkg/auth.ServerUnaryInterceptor(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xc01076c5c0, 0xc010735040, 0xc0107865a0, 0x3, 0x3, 0xf47d80)
	/home/falce/vchain/immudb/src/pkg/auth/serverinterceptors.go:59 +0xd9
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0x21, 0xc010735130, 0xc000220970, 0xc000220910)
	/home/falce/vchain/immudb/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.0/chain.go:34 +0x9c
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xc01076c5c0, 0xc010735040, 0xe03701, 0xc0107ce088, 0xc0002209f0, 0x14)
	/home/falce/vchain/immudb/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xad
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0x0, 0x14, 0x50, 0x48)
	/home/falce/vchain/immudb/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.0/chain.go:34 +0x9c
github.com/codenotary/immudb/pkg/server.(*uuidContext).UuidContextSetter(0xc00003e910, 0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xc01076c5c0, 0xc010735040, 0x50, 0xd93000, 0x1, ...)
	/home/falce/vchain/immudb/src/pkg/server/uuid.go:105 +0x1aa
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0xf42060, 0xc0107864e0, 0xd85900, 0xc010786510, 0xc01076c5c0, 0xc01076c5e0, 0xc000220b58, 0x4fd6b8, 0xd791a0, 0xc0107864e0)
	/home/falce/vchain/immudb/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.0/chain.go:39 +0x155
github.com/codenotary/immudb/pkg/api/schema._ImmuService_Backup_Handler(0xdef740, 0xc0003d25a0, 0xf42060, 0xc0107864e0, 0xc010653140, 0xc0003bca50, 0xf42060, 0xc0107864e0, 0x0, 0x0)
	/home/falce/vchain/immudb/src/pkg/api/schema/schema.pb.go:2882 +0x14b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000c21a0, 0xf4c4c0, 0xc00009f800, 0xc0106fc900, 0xc00026f200, 0x16d00c8, 0x0, 0x0, 0x0)
	/home/falce/vchain/immudb/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1082 +0x4fd
google.golang.org/grpc.(*Server).handleStream(0xc0000c21a0, 0xf4c4c0, 0xc00009f800, 0xc0106fc900, 0x0)
	/home/falce/vchain/immudb/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:1405 +0xd25
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0057bae30, 0xc0000c21a0, 0xf4c4c0, 0xc00009f800, 0xc0106fc900)
	/home/falce/vchain/immudb/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:746 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/falce/vchain/immudb/pkg/mod/google.golang.org/grpc@v1.29.1/server.go:744 +0xa1

```

**Comment by padurean** [comment]
@mmeloni Ok. That's unfortunate. I'd say we drop the _**online**_ mode, at least for `restore`, but maybe for both.
P.S. Let me know how you've tested as i didn't notice the error when i did the tests.

**Comment by mmeloni** [comment]
@padurean 
Backup online seems working, I made multiple tests on it and I'm not able to reproduce again but
restore --online crush DB while I'm running multiple client which write simultaneously on immugw.
To replicate use https://github.com/mmeloni/newman-parallel-run and try to restore when test is running.
It's up to you postpone online features

**Comment by padurean** [comment]
@mmeloni i've removed the _**online**_ backup and restore features (they are still available in a separate feature branch [feat/immuadmin-online-backup-restore](https://github.com/codenotary/immudb/commit/ae31f4429f0dde1e8bc63fd596ab5ce15ac65471)).
Hence only the _**offline**_ mode has been kept here.
If you don't have other remarks/change requests, please feel free to merge whenever you want.

**Pull request merged**